subsys/mgmt/mcumgr/grp/img_mgmt/src/zephyr_img_mgmt.c

```diff
@@ -614,30 +616,35 @@ int img_mgmt_upload_inspect(const struct img_mgmt_upload_req *req,
 			return IMG_MGMT_RET_RC_NO_FREE_SLOT;
 		}
 
-#if defined(CONFIG_MCUMGR_GRP_IMG_REJECT_DIRECT_XIP_MISMATCHED_SLOT)
-		if (hdr->ih_flags & IMAGE_F_ROM_FIXED) {
-			const struct flash_area *fa;
+		rc = flash_area_open(action->area_id, &fa);
+		if (rc) {
+			IMG_MGMT_UPLOAD_ACTION_SET_RC_RSN(action,
+				img_mgmt_err_str_flash_open_failed);
+			LOG_ERR("Failed to open flash area ID %u: %d", action->area_id, rc);
+			return IMG_MGMT_RET_RC_FLASH_OPEN_FAILED;
+		}
 
-			rc = flash_area_open(action->area_id, &fa);
-			if (rc) {
-				IMG_MGMT_UPLOAD_ACTION_SET_RC_RSN(action,
-					img_mgmt_err_str_flash_open_failed);
-				LOG_ERR("Failed to open flash area ID %u: %d", action->area_id,
-					rc);
-				return IMG_MGMT_RET_RC_FLASH_OPEN_FAILED;
-			}
+		/* Check that the area is of sufficient size to store the new image */
+		if (req->size > fa->fa_size) {
+			IMG_MGMT_UPLOAD_ACTION_SET_RC_RSN(action,
+				img_mgmt_err_str_image_too_large);
+			flash_area_close(fa);
+			LOG_ERR("Upload too large for slot: %u > %u", req->size, fa->fa_size);
+			return IMG_MGMT_RET_RC_INVALID_IMAGE_TOO_LARGE;
```

Comment:
This is good enough, although we probably need some boot API function to round down the size to actually available.
This will give us some problems because we are also able to upload already padded images.

Reply:
For the application via MCUmgr this should not be a problem because the status is in the primary image slot which will be running (assuming swap using move mode)